### ISSUE TITLE
`Task { [weak self] in ... }` loses its actor context when called in non-global actor method

### ISSUE THREAD

**Issue by inamiy** [apple/swift#62604]
**Description**

TLDR: https://twitter.com/inamiy/status/1603267958209331200

`Task { [weak self] in ... }` called in non-global actor's method seems to lose its actor context as follows:

```swift
// NOTE: Replacing with `@MainActor class` will compile.
actor A {
    var prop: Int = 0

    func foo() {
        // OK
        Task {
            self.prop = 1
        }

        // ERROR: Actor-isolated property 'prop' can not be mutated from a non-isolated context
        Task { [weak self] in
            self?.prop = 1
        }

        // ERROR: Actor-isolated property 'prop' can not be mutated from a non-isolated context
        Task { [weak self] in
            guard let self = self else { return }
            self.prop = 1
        }

        // Requires setter and `await`
        Task { [weak self] in
            await self?.setProp(1)
        }
    }

    func setProp(_ prop: Int) {
        self.prop = prop
    }
}

let a = A()
await a.foo()
```

**Steps to reproduce**

Try above code and notice compile error for non-global actor.

**Expected behavior**

As `@MainActor` works correctly, I _believe_ non-global actor code will also compile without error.

However, due to global actor having app's entire lifetime but not for non-global actor,
the same rule may not be applied, and having compiler error could be an expected error without needing any changes.

**Environment**
- Swift compiler version info: swift-driver version: 1.62.15 Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51) Target: arm64-apple-macosx13.0
- Xcode version info: Xcode 14.2 Build version 14C18

**Comment by ktoso** [comment]
Thanks for the report; I've not looked deeper yet, but perhaps this is more about losing isolation since the self is optional here which is kind of expected. We could perhaps gain it back after the guard... 

rdar://103427235

**Comment by ktoso** [comment]
Now tracked as rdar://100000802
